Duplicate Ways Check

savannahostrowski opened the pull request:
The Duplicate Ways check flags Ways in OpenStreetMap which are either partially or completely duplicated.

This pull request covers the logic, unit tests, documentation, and configuration for the Duplicate Ways check.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,109 @@
+    public static final String AREA_KEY = AreaTag.KEY;
+
+    // A map of segments and list of Edge identifiers
+    private final Map<Segment, Set<Long>> globalSegments = new HashMap<>();
```
Is some cases, this data structure has a potential to fill the memory, cause OOM exceptions. I'd suggest finding a less memory-intensive approach to find duplicate ways.

savannahostrowski commented:
I've refactored the logic for this check based on @mertemin's comments surrounding memory usage. Using a HashSet within the flag method scope(as opposed to the original global HashMap), I think this is more along the lines of what we're looking for. I've also added in another preliminary filtering conditional to ensure that we are only checking master Edges in the Duplicate Ways check.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+    public static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    public static final int ZERO_LENGTH = 0;
+    public static final String AREA_KEY = AreaTag.KEY;
```
These statics don't have to be exposed publicly from the class.

mertemin replied:
Using `AreaTag.KEY` might be cleaner than using `AREA_KEY`.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+                // Check to see that the Edge is a master Edge
+                && ((Edge) object).isMasterEdge()
+                // Check to see that the edge has not already been seen
+                && !this.isFlagged(((Edge) object).getMasterEdgeIdentifier())
```
If this condition is being checked, then `((Edge) object).isMasterEdge()` must be true, as well as `object instanceof Edge`. So you don't have to cast `object` here. Just get it's identifer and check if it is flagged or not.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+    {
+        return object instanceof Edge
+                // Check to see that the Edge is a master Edge
+                && ((Edge) object).isMasterEdge()
```
Please consider using [`Edge.isMasterEdgeIdentifier` method](https://github.com/osmlab/atlas/blob/1dbcbe5818d018f7c919b51d8ea82902ec1a5fa7/src/main/java/org/openstreetmap/atlas/geography/atlas/items/Edge.java#L30).

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+
+    public static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    public static final int ZERO_LENGTH = 0;
```
Please reuse [`Distance.ZERO`](https://github.com/osmlab/atlas/blob/1dbcbe5818d018f7c919b51d8ea82902ec1a5fa7/src/main/java/org/openstreetmap/atlas/utilities/scalars/Distance.java#L24).

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+            // If the Edge found in the bounds has an area tag or if the Edge has a length of 0
+            // Or if the Edge has already been flagged before then continue because we don't want to
+            // Flag area Edges or duplicate Nodes
+            if (edgeInBounds.getTags().containsKey(AREA_KEY)
```
`getTags` is fetching all the tags. Please rely on `getTag` method.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+            // Flag area Edges or duplicate Nodes
+            if (edgeInBounds.getTags().containsKey(AREA_KEY)
+                    || edgeInBounds.asPolyLine().length().equals(Distance.meters(ZERO_LENGTH))
+                    || this.isFlagged(edgeInBounds.getMasterEdgeIdentifier()))
```
`edgeInBounds` is verified that it is master edge on line 73. No need to call `getMasterEdgeIdentifier`.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+            }
+
+            // If the Set of Edges does not contain the Edge found in the bounds
+            if (!allEdgeSegments.containsAll(edgeInBounds.asPolyLine().segments()))
```
Please consider using [`PolyLine.intersects` method](https://github.com/osmlab/atlas/blob/c0965af0fd1563844fff505a2a80cc79b95f655e/src/main/java/org/openstreetmap/atlas/geography/PolyLine.java#L592) to check for an overlap. You can also utilize [`PolyLine.intersections` method](https://github.com/osmlab/atlas/blob/c0965af0fd1563844fff505a2a80cc79b95f655e/src/main/java/org/openstreetmap/atlas/geography/PolyLine.java#L570) to get overlapping locations.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+            {
+                this.markAsFlagged(edgeInBounds.getMasterEdgeIdentifier());
+                return Optional.of(this.createFlag(edgeInBounds,
+                        this.getLocalizedInstruction(0, edgeInBounds.getOsmIdentifier())));
```
A full overlap has the same instruction. I'd suggest not worrying about the full overlap. Just flag whenever an overlap is found. Atlas has edges vs OSM has ways. Edges by definition are partial OSM ways, so full overlap or not, you'll almost always flag partial OSM ways anyways.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,115 @@
+        // Get Edges which are contained by or intersect the bounds, and then filter
+        // Out the non-master Edges as the bounds Edges are not guaranteed to be uni-directional
+        final Iterable<Edge> edgesInBounds = Iterables
+                .stream(edge.getAtlas().edgesIntersecting(bounds)).filter(Edge::isMasterEdge);
```
You can possibly say `final Iterable<Edge> edgesInBounds = edge.getAtlas().edgesIntersecting(bounds, Edge::isMasterEdge)`. `edgesIntersecting` optionally accepts a `Predicate`.

savannahostrowski commented:
Additionally, for my own knowledge, was the concern here mostly just to do with the use of the HashMap to store Segments and associated Edge IDs as a global variable?  Or is the concern more specific to how Atlas Checks are being run (is it being run on a country-by-country/region-by-region basis or is the whole world's data being processed in one fell swoop? 😄 )

mertemin commented:
@savannahostrowski Atlas-checks is by design flexible to accept any given part of the world to execute (or the whole world for that matter). The concern with accumulating state and using it as you go validate more features is that results will be determined by the order of features processed and memory footprint of a such solution could be big (based on the area of interest). Atlas could be utilized and possibly overlapping edges could be retrieved without additional memory tracking.

Here is an example:
edge-x (A B C D)
edge-y (H I K L)
edge-z (A B C)
edge-t (C D)

If you process x, y, z and finally t, then you'd flag z and t for overlap. If you process, t, z, y and finally x, then x will be flagged only.

savannahostrowski commented:
Addressed @mertemin comments regarding reuse of constants, and fine tuning preliminary filtering. Opted to use [.overlapsShapeOf()](https://github.com/osmlab/atlas/blob/c0965af0fd1563844fff505a2a80cc79b95f655e/src/main/java/org/openstreetmap/atlas/geography/PolyLine.java#L830) to detect duplicate Ways. Let me know if this still isn't sufficient!

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,99 @@
+
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    private static final Distance ZERO_LENGTH = Distance.ZERO;
```
Consider using `ZERO_DISTANCE` to be consistent with the type.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,99 @@
+                // Check to see that the edge is car navigable
+                && HighwayTag.isCarNavigableHighway(object)
+                // The edge is not part of an area
+                && !object.getTags().containsKey(AreaTag.KEY);
```
Consider using `object.getTag(AreaTag.KEY)` to be consistent (see line #77).

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java:
```diff
@@ -0,0 +1,99 @@
+            // Checks that either the edgeInBounds PolyLine overlaps the edgePoly or
+            // That the edgePoly overlaps the edgeInBoundsPoly
+            if (edgeInBoundsPoly.overlapsShapeOf(edgePoly)
+                    || edgePoly.overlapsShapeOf(edgeInBoundsPoly))
```
`overlapsShapeOf` is comparing individual segments. If there is overlap, then that would be found with the first call. If there is no overlap, then it wouldn't be found in both overlap checks. I'd suggest dropping one of the overlap calls, as it seems redundant.
